src/main/java/fi/fmi/avi/model/taf/immutable/TAFImpl.java

```diff
@@ -178,6 +181,257 @@ public Builder withCompleteForecastTimes(final ZonedDateTime reference) {
+        }
+
+        /**
+         * Sets the link to another (referred) report used for cancellation and amendment messages.
+         * <p>
+         * Also sets the {@link #setCancelledReportValidPeriod(PartialOrCompleteTimePeriod)} using the {@link TAFReference#getValidityTime()} value if
+         * {@link #isCancelMessage()} equals true.
+         *
+         * @param referredReport the reference
+         * @return the builder
+         */
+        @Override
+        @JsonDeserialize(as = TAFReferenceImpl.class)
+        public Builder setReferredReport(final TAFReference referredReport) {
+            super.setReferredReport(referredReport);
+            if (this.isCancelMessage() && referredReport.getValidityTime().isPresent()) {
+                super.setCancelledReportValidPeriod(referredReport.getValidityTime().get());
+            }
+            return this;
+        }
+
+        /**
+         * Sets the cancellation status of this message.
+         * If <code>cancel == true</code> also calls {@link #setReportStatus(ReportStatus)} with
+         * {@link fi.fmi.avi.model.AviationCodeListUser.TAFStatus#CANCELLATION}. Otherwise resets the <code>status</code> value based on the
+         * {@link #getReportStatus()} value.
+         *
+         * @param cancel true to set as cancellation, false to unset
+         * @return the builder
+         */
+        @Override
+        public Builder setCancelMessage(final boolean cancel) {
+            super.setCancelMessage(cancel);
+            if (cancel) {
+                super.setStatus(TAFStatus.CANCELLATION);
+            } else {
+                if (this.getReportStatus().isPresent()) {
+                    this.setReportStatus(this.getReportStatus()); // takes care of setting the status based on reportStatus
+                }
+            }
+            if (cancel && this.getReferredReport().isPresent()) {
+                this.setCancelledReportValidPeriod(this.getReferredReport().get().getValidityTime());
+            }
+            return this;
+        }
+
+        /**
+         * Sets the cancellation status of this message.
+         * If <code>missing == true</code> also calls {@link #setReportStatus(AviationWeatherMessage.ReportStatus)} with
+         * {@link fi.fmi.avi.model.AviationCodeListUser.TAFStatus#MISSING}. Otherwise resets the <code>status</code> value based on the
+         * {@link #getReportStatus()} value.
+         *
+         * @param missing true to set as missing message, false to unset
+         * @return the builder
+         */
+        @Override
+        public Builder setMissingMessage(final boolean missing) {
+            super.setMissingMessage(missing);
+            if (missing) {
+                super.setStatus(TAFStatus.MISSING);
+            } else {
+                if (this.getReportStatus().isPresent()) {
+                    this.setReportStatus(this.getReportStatus()); // takes care of setting the status based on reportStatus
```

Comment:
Though the whole `if` block can be removed if `getStatus()` value is [not stored](https://github.com/fmidev/fmi-avi-messageconverter/pull/68#discussion_r523933303).

Reply:
Yeah, this looks ugly, but does actually have an effect since setReportStatus has the desired side-effect of updating the  the TAFStatus. I'll try to get rid of this entirely by determining the status value of-the-fly as suggested